src/samVisualize/graphGeneration/cfnTemplateGraphGenerator.ts

```diff
@@ -0,0 +1,138 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { Graph, GraphObject } from './graph'
+import { yamlParse } from 'yaml-cfn'
+import { LinkTypes } from './linkTypes'
+import { isString, isObjectLike } from 'lodash'
+
+/**
+ * @param inputYaml A string representing a YAML template
+ * @returns A JavaScript object corresponding to the given YAML string
+ */
+function yamlStringToObject(inputYaml: string): Record<string, unknown> {
+    let templateAsObject = {}
+    try {
+        templateAsObject = yamlParse(inputYaml)
+    } catch (err) {
+        console.error(err)
+    }
+    return templateAsObject
+}
+
+/**
+ * @description Takes a string with embedded resource names within `${}`, and extracts each resource.
+ * See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html#w2ab1c33c28c59c11b6
+ * @example 'xxx${resource1}yyyzzz${resource2}abc' => ['resource1', 'resource2']
+ * @param value A string containing embedded resources
+ * @returns A list of strings corresponding to the embedded resource names. If no resource names are found, returns an empty list.
+ */
+function extractSubstitution(value: string): Array<string> {
+    const names: Array<string> | null = value.match(/\${[^}]+}/g)
+    return names ? names.map((str: string) => str.replace(/(\$|}|{)/g, '')) : []
+}
+
+/**
+ * Returns whether or not a given entity can be traversed
+ */
+function isTraversable(entity: any): boolean {
+    // Ensure it's of type object (and not null or undefined)
+    return isObjectLike(entity)
+}
+
+/**
+ * Recursively traverses (DFS) an object representation of a yaml template resource, adding links to the graph where found
+ * @param graph The Graph in which links are created. It is modified, due to traverse being recursive. But this Graph is not visible from outside the module, so there are no external side effects
+ * @param currentObject The object wich is currently being searched for link keys
+ * @param parentNodeName The name of the node representing the current parent node, which any found links will be added to.
+ */
+function traverse(graph: Graph, currentObject: Record<string, any>, parentNodeName: string): void {
```

Comment:
Quite a bit of code in this function is deeply nested, I'd suggest moving some of the bigger case blocks into their own functions.

Reply:
Since a function that would take the place of a case block would have many parameters, and likely reduce clarity, I was thinking of an alternative to reducing nesting. Instead of checking for traverse-ability before each recursive call, we can just check at the beginning of the `traverse` call. That way, we can eliminate every `if(isTraversable(x)` block. The resulting case blocks will have at most an if statement with a for loop.

Reply:
Yeah that will help with the nesting/readability. This function doesn't have any super complicated logic going on so the nesting isn't a huge deal in my opinion.